### PR TITLE
[WIP] Cloud Hypervisor Unified Binary

### PR DESCRIPTION
**Pull request by thesantatitan**
This is an initial proof-of-concept PR for the issue #4079.

- For a lot re-exported types, I have defined my own structs, which convert to hypervisor-specific structs.
- For the remaining types I have created a struct which has a private enum containing a variant for each hypervisor as suggested      [here](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4079#issuecomment-1123520869)
- I have also added a global static variable to store the hypervisor type. This was required for some runtime decisions, where we need to figure out which hypervisor is being used, for example in the function boot_msr_entries()

### CODE REVIEW

**Comment by liuw** [comment]
Thanks for posting this PR.

> This is an initial proof-of-concept PR for the issue #4079.
> 
> * For a lot re-exported types, I have defined my own structs, which convert to hypervisor-specific structs.
> * For the remaining types I have created a struct which has a private enum containing a vraint for each hypervisor as suggested      [here](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4079#issuecomment-1123520869)
> * I have also added a global static variable to store the hypervisor type. This was required for some runtime decsions, where we need to figure out which hypervisor is being used, for example in the function boot_msr_entries()

I would like to see the global variable eliminated. It should be doable but requires a bit more refactoring for the common code.

I would like to ask people's opinions on where to put the new structures. I think we can just use `hypervisor/arch/x86/mod.rs` for simplicity. @rbradford @sboeuf your opinions?

The new structures won't need to have `repr(C)` or any padding fields -- they don't need to interface with C code. As is the code probably breaks migration because it skips all CPUID fields during serialization and deserialization. The CPUID flexible array can perhaps just be replaced with a vector. That removes code complexity and will likely fix migration (once it gets properly serialized/deserialized).

To be clear while I want to point out the deficiencies in the code (mostly for my own references), this is by no mean criticism towards Dev's work. Dev came in without knowing Rust or virtualization or Cloud Hypervisor at all, yet still produced a working prototype within a few weeks, that itself was quite a feat.

P.S. Since Dev's internship ends today, either I or Jinank will take over the patches.

**Comment by rbradford** [comment]
@liuw Perhaps worth adding "WIP" to the title of the PR to skip Jenkins runs until the code reliably builds?

**Comment by liuw** [comment]
> > * I have also added a global static variable to store the hypervisor type. This was required for some runtime decsions, where we need to figure out which hypervisor is being used, for example in the function boot_msr_entries()
> 
> I would like to see the global variable eliminated. It should be doable but requires a bit more refactoring for the common code.

I reworked `boot_msr_entries` to not require this global variable. I think I should be able to eliminate all dependencies of such a variable in the `hypervisor` crate.

I thought the global variable was only used in the hypervisor crate. But I obviously overlooked that the same variable is also needed by the `vmm` crate. That's fair. I myself has some patches to keep that bit information around. I think that's a fair thing to do.

**Comment by liuw** [comment]
> @liuw Perhaps worth adding "WIP" to the title of the PR to skip Jenkins runs until the code reliably builds?

Just did that.

I think the series needs quite a bit of rework to meet upstream's standard -- not every single patch builds on its own.

**Comment by rbradford** [comment]
> > @liuw Perhaps worth adding "WIP" to the title of the PR to skip Jenkins runs until the code reliably builds?
> 
> Just did that.
> 
> I think the series needs quite a bit of rework to meet upstream's standard -- not every single patch builds on its own.

I was wondering it if makes sense to split it up into smaller PRs?

**Comment by liuw** [comment]
> > > @liuw Perhaps worth adding "WIP" to the title of the PR to skip Jenkins runs until the code reliably builds?
> > 
> > 
> > Just did that.
> > I think the series needs quite a bit of rework to meet upstream's standard -- not every single patch builds on its own.
> 
> I was wondering it if makes sense to split it up into smaller PRs?

I will see what I can do.

Right now we need to decide where to put the code. The transition to common structures can be perhaps done piece by piece.

File: src/main.rs [hunk 1]
```diff
@@ -528,8 +529,18 @@ fn start_vmm(cmd_arguments: ArgMatches) -> Result<Option<String>, Error> {
     }
 
     event!("vmm", "starting");
-
-    let hypervisor = hypervisor::new().map_err(Error::CreateHypervisor)?;
+    
+    let hypervisor_type = if let Some(x) = cmd_arguments.value_of("hypervisor") {
```
[the remark this answers is not in the document]
Comment (liuw) [reply src/main.rs:533]: I have some patches for that already. They are not in this series though.

**Comment by liuw** [comment]
This project is done. See https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4345.

**Pull request closed without merging**